resources/asciidoctor/lib/edit_me/extension.rb

```diff
@@ -28,50 +28,73 @@ def process(document)
       url = url[0..-2] if url.end_with? '/'
       edit_urls << { toplevel: toplevel, url: url }
     end
-    edit_urls.reverse!
+    # Prefer the longest matching edit url
+    edit_urls = edit_urls.sort_by { |e| [-e[:toplevel].length, e[:toplevel]] }
     document.attributes['edit_urls'] = edit_urls
     super
   end
 
   def process_block(block)
     return unless %i[preamble section floating_title].include? block.context
 
-    def block.title
+    block.extend WithEditLink
+    if block.context == :preamble
+      def block.source_path
+        # source_location.path doesn't work for relative includes outside of
+        # the base_dir which we use when we build books from many repos.
+        document.source_location.file
+      end
+    else
+      def block.source_path
+        # source_location.path doesn't work for relative includes outside of
+        # the base_dir which we use when we build books from many repos.
+        source_location.file
+      end
+    end
+  end
+
+  ##
+  # Extension to blocks that need an "edit me" link.
+  module WithEditLink
+    def title
+      url = edit_url
+      return super unless url
+
+      "#{super}<ulink role=\"edit_me\" url=\"#{edit_url}\">Edit me</ulink>"
+    end
+
+    def edit_url
+      if @document.attributes['respect_edit_url_overrides']
+        url = @document.attributes['edit_url']
+        return false if url == ''
+        return url if url
+      end
+
+      edit_url_by_path
+    end
+
```

Comment:
The `if` block will return the last evaluated statement so you could also write this:
```ruby
def edit_url
  if @document.attributes['respect_edit_url_overrides']
    url = @document.attributes['edit_url']
    return false if url == ''
    url
  end || edit_url_by_path
end
```